File: src/util/PuzzleSolver.ts
import { PuzzleInputReader } from './PuzzleInputReader';

/**
 * Base class for solving puzzles.
 */
export abstract class PuzzleSolver {
  private readonly puzzleNumber: number;

  private _inputReader: PuzzleInputReader;

  get inputReader(): PuzzleInputReader {
    return this._inputReader;
  }

  constructor(puzzleNumber: number) {
    this._inputReader = new PuzzleInputReader(puzzleNumber);
    this.puzzleNumber = puzzleNumber;
  }

  protected loadPuzzleTestData(testNr: number): void {
    this._inputReader = new PuzzleInputReader(this.puzzleNumber, testNr);
  }

  /**
   * Prints the solution in a unified way to the console.
   *
   * @param solution Solution to print.
   * @param part Part of the puzzle the solution belongs to.
   */
  protected printSolution(solution: number, part: 'A' | 'B'): void {
    console.log(`Solution ${part}: ${solution}`);
  }

  /**
   * Solves the given puzzle.
   */
  abstract solve(): void;
}

File: kotlin/common/src/main/kotlin/PuzzleSolver.kt
import kotlin.system.measureTimeMillis

abstract class PuzzleSolver(private val puzzleNumber: Int) {
    private var inputReader: PuzzleInputReader = PuzzleInputReader(puzzleNumber)

    abstract fun solvePartA(): Number

    abstract fun solvePartB(): Number

    fun solve() {
        solvePart(PuzzlePart.A, this::solvePartA)
        solvePart(PuzzlePart.B, this::solvePartB)
    }

    @Suppress("unused")
    protected fun getPuzzleInput(): String = inputReader.getPuzzleInput()

    protected fun getPuzzleLines(): List<String> = inputReader.getPuzzleInputSplitByLines()

    @Suppress("unused")
    protected fun loadPuzzleTestData(testNo: Int) {
        inputReader = PuzzleInputReader(puzzleNumber, testNo)
    }

    private fun solvePart(part: PuzzlePart, action: () -> Number) {
        try {
            var solution: Number
            val time = measureTimeMillis { solution = action() }
            printSolution(solution, part, time)
        } catch (e: NotImplementedError) {
            println("Part B is not yet implemented")
        }
    }

    private fun printSolution(solution: Number, part: PuzzlePart, time: Long) {
        println("Part $part: $solution ($time ms)")
    }
}

enum class PuzzlePart {
    A, B
}
